test/Interop/Cxx/foreign-reference/Inputs/inheritance.h

```diff
@@ -68,3 +70,188 @@ DerivedOutOfOrder : public BaseT, public DerivedWithVirtualDestructor {
+__attribute__((swift_attr("release:baseRelease2"))) BaseRef2 {};
+void baseRetain2(BaseRef2 *v) {}
+void baseRelease2(BaseRef2 *v) {}
+BaseRef2 *returnBaseRef2() { // expected-warning {{'returnBaseRef2' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it is returning a SWIFT_SHARED_REFERENCE}}
+  return new BaseRef2();
+};
+
+struct __attribute__((swift_attr("import_reference")))
+__attribute__((swift_attr("retain:baseRetain1")))
+__attribute__((swift_attr("release:baseRelease1"))) BaseRef3 : BaseRef1 {};
+BaseRef3 *returnBaseRef3() {
+  return new BaseRef3();
+};
+
+struct DerivedFromBaseRef1AndBaseRef2 : BaseRef1, // expected-warning {{unable to infer 'SWIFT_SHARED_REFERENCE' for 'DerivedFromBaseRef1AndBaseRef2', although one of its C++ inheritance parents is marked as 'SWIFT_SHARED_REFERENCE'}}
+                                        BaseRef2 {
+};
+DerivedFromBaseRef1AndBaseRef2 *returnDerivedFromBaseRef1AndBaseRef2() {
+  return new DerivedFromBaseRef1AndBaseRef2();
+};
+
+struct DerivedFromBaseRef3 : BaseRef3 {};
+DerivedFromBaseRef3 *
+returnDerivedFromBaseRef3() { // DOUBT: where is the warning??
+  return new DerivedFromBaseRef3();
+};
+
+struct __attribute__((swift_attr("import_reference")))
+__attribute__((swift_attr("retain:bRetain1")))
+__attribute__((swift_attr("release:bRelease1"))) B1 {};
+void bRetain1(B1 *v) {}
+void bRelease1(B1 *v) {}
+B1 *returnB1() { // expected-warning {{'returnB1' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it is returning a SWIFT_SHARED_REFERENCE}}
+  return new B1();
+};
+
+struct __attribute__((swift_attr("import_reference")))
+__attribute__((swift_attr("retain:bRetain2")))
+__attribute__((swift_attr("release:bRelease2"))) B2 {};
+void bRetain2(B2 *v) {}
+void bRelease2(B2 *v) {}
+B2 *returnB2() { // expected-warning {{'returnB2' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it is returning a SWIFT_SHARED_REFERENCE}}
+  return new B2();
+};
+
+struct __attribute__((swift_attr("import_reference")))
+__attribute__((swift_attr("retain:bRetain3")))
+__attribute__((swift_attr("release:bRelease3"))) B3 {};
+void bRetain3(B3 *v) {}
+void bRelease3(B3 *v) {}
+B3 *returnB3() { // expected-warning {{'returnB3' should be annotated with either SWIFT_RETURNS_RETAINED or SWIFT_RETURNS_UNRETAINED as it is returning a SWIFT_SHARED_REFERENCE}}
+  return new B3();
+};
+
+struct D : B1, // expected-warning {{unable to infer 'SWIFT_SHARED_REFERENCE' for 'D', although one of its C++ inheritance parents is marked as 'SWIFT_SHARED_REFERENCE'}}
+           B2,
+           B3 {};
+D *returnD() { return new D(); };
+
+// To check for x, y, x kind of pattern in parent's retain/release function name
+// when infering SWIFT_SHARED_REFERENCE
+struct __attribute__((swift_attr("import_reference")))
+__attribute__((swift_attr("retain:retain1")))
+__attribute__((swift_attr("release:release1"))) B1_ {};
```

Comment:
Using very similar names over and over again like `B1` and `B1_` can get confusing. If you put the different test cases in different namespaces that would make it clear what classes should be read together, would give a nice description of what are you testing (in the name of the namespace), and you could reuse the same type names in multiple test cases.